src/core/packages/context-browser-internal/src/render_context_service.tsx

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import React from 'react';
+import useObservable from 'react-use/lib/useObservable';
+
+import { BehaviorSubject } from 'rxjs';
+
+import { EuiLoadingSpinner } from '@elastic/eui';
+import type { AnalyticsServiceStart } from '@kbn/core-analytics-browser';
+import type { ExecutionContextStart } from '@kbn/core-execution-context-browser';
+import type { I18nStart } from '@kbn/core-i18n-browser';
+import type { RenderContextService as IRenderContextService } from '@kbn/core-render-context-browser';
+import type { UserProfileService } from '@kbn/core-user-profile-browser';
+import type { ThemeServiceStart } from '@kbn/react-kibana-context-common';
+import { KibanaRenderContextProvider } from '@kbn/react-kibana-context-render';
+
+/**
+ * Startup services needed for rendering fully-featured React nodes in Kibana
+ */
+interface RenderingServiceDeps {
+  analytics: AnalyticsServiceStart;
+  executionContext: ExecutionContextStart;
+  i18n: I18nStart;
+  theme: ThemeServiceStart;
+  userProfile: UserProfileService;
+}
+
+/**
+ * Carries startup services in internal state and can add the necessary context for out-of-current React
+ * rendering, such as using `ReactDOM.render()`.
+ *
+ * Usage:
+ *
+ * export const renderApp = (core: CoreStart, props: Props, element: HTMLElement) => {
+ *   const { rendering } = core; // obtain instance of RenderContextService from CoreStart
+ *   ReactDOM.render(rendering.addContext(<MyApp {...props} />), element);
+ *   return () => ReactDOM.unmountComponentAtNode(element);
+ * };
+ *
+ */
+export class RenderContextService implements IRenderContextService {
+  private deps = new BehaviorSubject<RenderingServiceDeps | null>(null);
+
+  start(deps: RenderingServiceDeps) {
+    this.deps.next(deps);
+    return this;
+  }
+
+  public addContext(element: React.ReactNode): React.ReactElement<string> {
+    const Component: React.FC = () => {
+      const deps = useObservable(this.deps, null);
```

Comment:
I think we should avoid this if possible because now we alway get 2 renders: 1st with null and then with deps. And this seem to be pretty important place where we should really try to optimize re-renders. 

Wouldn't it be safe to say that when Component is rendered, deps are already available? 

so just? 

```suggestion
      const deps = this.deps
```

or at least

```suggestion
      const deps = useObservable(this.deps, this.deps.getValue());
```

Reply:
> Wouldn't it be safe to say that when Component is rendered, deps are already available?

@Dosant 
Originally I had the code acting more synchronously and would throw an error if deps were not available in `addContext`. When I updated the mount of the Feedback button in `x-pack/platform/plugins/shared/serverless/public/plugin.tsx` I was seeing that error get thrown, so I rewrote this class to use BehaviorSubject. I'm not able to repro the error now; it looks like an inconsistent race condition, or perhaps there was something wrong with my setup earlier.

I still assume that plugins' start methods could be called before the CoreStart services are completely settled. I think that setting `this.deps.getValue()` for an initial value in `useObservable` is safer and it does avoid unnecessary re-renders. Thanks!